rfcs/20210119-determinism.md

```diff
@@ -0,0 +1,126 @@
+
+Certain random ops, such as `tf.image.sample_distorted_bounding_box` and `tf.nn.fractional_max_pool`, ignore the global seed if a seed is not explicitly passed. For such ops, setting the global seed is not enough to avoid the error, so users must pass a seed directly to the op.
+
+As for TensorFlow 2 random number generation, `tf.random.Generator.from_non_deterministic_state` will raise an error if called when determinism is enabled. In such cases, users should check if determinism is enabled and if so, use a different generator from a deterministic source. `tf.random.get_global_generator` implicitly calls `from_non_deterministic_state` if no global generator is set, and so will also raise an error if a global generator is not set with `tf.random.set_global_generator`.
+
+Stateless random functions, such as `tf.random.stateless_normal`, are always deterministic and so will never raise determinism-related errors.
+
+No error will be raised if a random op or generator is run before determinism is enabled (as is true for any other op), so users should take care to enable determinism before running any random ops or generators.
+
+### Testing plan
+
+We must ensure that every op will either run deterministically or raise an error if `enable_deterministic_ops` has been called. In order to do this, we must test for determinism, both to check we don't accidentally miss any nondeterministic ops and to ensure any previously-deterministic ops are not accidentally made nondeterministic in the future. To accomplish this, we will have the following tests:
+
+1. We will add tests to several of the [official models](https://github.com/tensorflow/models/tree/master/official) to ensure they run deterministically. In particular, each test will train a model for several steps, then retrain it from scratch several times. The final weights after training will be asserted to be the same each time. This tests not only the `enable_deterministic_ops` API but that the entire model is deterministic. This only tests ops that the official models use.
+
+2. We will add a special mode to TensorFlow where every time a non-stateful op is run, TensorFlow will rerun the op several times and assert the outputs are the same each time. We will then run the TensorFlow unit tests with this mode as part of the nightly tests. Doing so ensures that for each op that is run as part of a unit test, it will be tested for determinism.
+
+3. When adding determinism to an op which previously was nondeterministic, an explicit unit test will be added that checks for determinism. This is slightly redundant with the special mode described above, but the explicit unit test can be part of the presubmit tests instead of the nightly tests, and can test on inputs that are very likely to demonstrate nondeterminism if it exists.
```

Comment:
I don't think directed testing is redundant at all. The automated testing approach, described previously, might (maybe) capture some small subset of issues, at potentially great additional compute cost, but it would miss most of the nondeterminism that we're trying to defend against.

A potentially more efficient and cost-effective way of defending against regressions might be to automatically flag PRs that contain suspect code, such as the use of CUDA atomic operations or sharding processes across CPU threads; then review those and/or require directed determinism tests (or `tf.errors.UnimplementedError` tests) for them.

Reply:
It's true, though, that once an op is proven to be thoroughly deterministic using a human-developed directed test, it's very unlikely that the op will somehow start functioning nondeterministically again. So the non-automated, directed tests do primarily provide evidence that a previously nondeterministic op has been made to operate deterministically.

Reply:
As you mentioned, it's unlikely the special mode would catch errors for an op which has an explicit unit tests, so there is *slight* redundancy. But I rephrased the sentence to avoid the word "redundancy".

I don't think it is worth it to flag PRs which have potentially unsafe constructs like atomics, since I think the effort to set that up will be high and there will be false positives. Others may disagree. /CC @sanjoy @pkanwar23